Cassandra 15229 trunk

jasonstack opened the pull request:
* Initiate 2 buffer pool instances, one for chunk cache (default 512mb), one for network (default 128mb).
* Improve buffer pool metrics to track: overflow allocation, buffer pool size in used.
* Allow recirculate partially freed chunk to improve cache utilization due to chunk cache holding buffer for arbitrary time period.

Fallout: [link](https://fallout.sjc.dsinternal.org/performance/zhao.yang@datastax.com/report/a978bfd6-5dae-49ea-984c-061301d496cf)

Circle-CI: [link](https://circleci.com/workflow-run/096afbe1-ec99-4d5f-bdaa-06f538b8280f)

jasonstack committed:
CASSANDRA-15229: Segregate Network and Chunk Cache BufferPools and
Recirculate Partially Freed Chunks.

* initiate multiple buffer pool for different lifespan and usages
 - Chunk Cache Buffer Pool - conf.file_cache_size_in_mb=512mb
 - Networking Buffer Pool - conf.temporary_cache_size_in_mb=128mb

* Add overflowSize and usedSize to buffer pool metrics

* re-circulate buffer pool Chunk for ChunkCache whenever it has free space, even thoughput it may not be able to allocate due to fragmentation

patch by Zhao Yang; reviewed by Caleb Rackliffe and Aleksey Yeschenko
for CASSANDRA-15229

jasonstack committed:
TEST ONLY - DO NOT MERGE

maedhroz commented on src/java/org/apache/cassandra/utils/memory/BufferPool.java:
```diff
@@ -68,23 +112,40 @@
+    private final LongAdder overflowMemoryUsage = new LongAdder();
+
+    /**
+     * SIze of buffer being used in bytes, including pooled buffer and unpooled buffer.
```
```suggestion
     * Size of buffer being used in bytes, including pooled buffer and unpooled buffer.
```

maedhroz commented on src/java/org/apache/cassandra/metrics/BufferPoolMetrics.java:
```diff
@@ -25,24 +25,47 @@
 
-    public BufferPoolMetrics()
+    /** Total size, in bytes, of active buffered being used from the pool currently + overflow */
+    public final Gauge<Long> usedSize;
```
@jasonstack So if there were to be a hole in the bookkeeping logic, one way we might know is that `usedSize` > `size`?

jasonstack replied:
if it's over-allocated, then `useSize` > `size`.

maedhroz commented on src/java/org/apache/cassandra/net/AsyncStreamingOutputPlus.java:
```diff
@@ -53,6 +54,8 @@
 {
     private static final Logger logger = LoggerFactory.getLogger(AsyncStreamingOutputPlus.class);
 
+    private static final BufferPool bufferPool = BufferPoolManager.temporary();
```
nit: If we ever want to mock out the pool for testing, we could make it an instance field, pass into the constructor, etc. (I suppose we don't carry around so many references this way.)

jasonstack replied:
I will change it to instance field.. we can mock it when tests need it.

maedhroz commented on test/unit/org/apache/cassandra/metrics/BufferPoolMetricsTest.java:
```diff
@@ -51,21 +52,21 @@ public static void setup() throws ConfigurationException
     @Test
     public void testMetricsSize()
     {
         // basically want to test changes in the metric being reported as the buffer pool grows - starts at zero
```
@jasonstack Should we also test the state of `usedSize` and `overflowSize` here? (We could have a separate test too, but piggybacking on this one seems okay too...)

jasonstack replied:
+1

maedhroz commented on src/java/org/apache/cassandra/utils/memory/PermanentBufferPool.java:
```diff
@@ -0,0 +1,31 @@
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.PooledByteBufAllocator;
+
+public class PermanentBufferPool
```
nit: unused class?

jasonstack replied:
removed..

maedhroz commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -302,6 +302,8 @@
     private static boolean isClientMode = false;
     private static Supplier<Config> overrideLoadConfig = null;
 
+    public Integer temporary_cache_size_in_mb;
```
We'll have to include a short section on this in `cassandra.yaml`, correct? Also, I'm guessing we'll have to make a couple minor changes to the description of `file_cache_size_in_mb`. For example...

```
32MB of this are reserved for pooling buffers, the rest is used as an
cache that holds uncompressed sstable chunks.
```

(Also there's a typo there, should be "a" not "an")

jasonstack replied:
+1

maedhroz commented on src/java/org/apache/cassandra/utils/memory/BufferPoolManager.java:
```diff
@@ -0,0 +1,78 @@
+
+import static org.apache.cassandra.utils.FBUtilities.prettyPrintMemory;
+
+public class BufferPoolManager
```
nit: If you're looking for a shorter name, `BufferPools` would probably be ok. (Think about `Executor` and `Executors`...)

iamaleksey replied:
Was going to suggest the same actually, as nothing is actually being managed here, just references held. There is a lot of precedent in the codebase for <Something>s class naming scheme. `InboundMessageHandlers` recently-ish, for example.

jasonstack replied:
good idea

maedhroz commented on src/java/org/apache/cassandra/cache/ChunkCache.java:
```diff
@@ -45,6 +46,8 @@
     private static boolean enabled = cacheSize > 0;
     public static final ChunkCache instance = enabled ? new ChunkCache() : null;
 
+    private static final BufferPool bufferPool = BufferPoolManager.permanent();
```
I was going to say it might be nice to just pass the `BufferPool` to the `ChunkCache` constructor, given it's already a singleton, and you could avoid accessing the manage entirely if `enabled == false`, but you might also have to access it from `Buffer`, which is static. (I guess you could still access it through `instance`, so maybe that's fine.) We might not care about testing ChunkCache w/ a custom `BufferPool` right this second, but getting in the habit of injecting dependencies is probably a good thing for the project.

jasonstack replied:
moved `bufferPool` into instance variable of `ChunkCache` and converted `Buffer` as instance class so it can access `bufferPool`

maedhroz commented on src/java/org/apache/cassandra/hints/ChecksummedDataInput.java:
```diff
@@ -43,6 +44,8 @@
  */
 public class ChecksummedDataInput extends RebufferingInputStream
 {
+    private static final BufferPool bufferPool = BufferPoolManager.permanent();
```
@jasonstack Possibly naïve question, but why do we use the buffer pool when we're reading hints/dispatching them to other nodes, etc.?

CC @iamaleksey

iamaleksey replied:
It's been used for a while now. We need a fixed size buffer here, and it makes sense to reuse an existing pool instead of implementing an ad-hoc one. That said, this particular use case would probably be just fine w/ allocating a buffer when needed instead of pooling at all.

jasonstack replied:
changed to allocating new buffer instead of pooled buffer.

maedhroz replied:
@jasonstack The only thing, having looked at this a bit more, is that `CompressedChecksummedDataInput` still uses the pool. Is it possible for `readBuffer()` to put the initial buffer, which wasn't _taken_ from the pool, back into the pool?

maedhroz commented on src/java/org/apache/cassandra/utils/memory/BufferPool.java:
```diff
@@ -68,23 +112,40 @@
+    protected final String name;
+    protected final BufferPoolMetrics metrics;
+    private long memoryUsageThreshold;
+    private String readableMemoryUsageThreshold; // FIXME
```
@jasonstack What does this FIXME entail?

jasonstack replied:
a left over during rebase.. removed now

maedhroz commented on src/java/org/apache/cassandra/utils/memory/BufferPoolManager.java:
```diff
@@ -0,0 +1,78 @@
+    {
+        logger.info("Global buffer pool limit is {} for {} and {} for {}",
+                    prettyPrintMemory(FILE_MEMORY_USAGE_THRESHOLD), PERMANENT.name,
+                    prettyPrintMemory(TEMPORARY_MEMORY_USAGE_THRESHOLD), TEMPORARY.name);
```
nit: It might be nice to have these on separate lines so we can grep them out of logs separately.

jasonstack replied:
+1

maedhroz commented on src/java/org/apache/cassandra/utils/memory/BufferPool.java:
```diff
@@ -68,23 +112,40 @@
+
+    protected final String name;
+    protected final BufferPoolMetrics metrics;
+    private long memoryUsageThreshold;
```
Now that `memoryUsageThreshold` is non-final and non-volatile, are there any risks around visibility between the thread that creates the pool and the threads that access it? We might be accidentally safe because of other locking, but I'd at least want to make sure we briefly document that...

maedhroz replied:
(...or we could just make it `volatile`)

maedhroz replied:
Same w/ `debug` and `readableMemoryUsageThreshold`.

jasonstack replied:
+1 to `volatile`

maedhroz commented on src/java/org/apache/cassandra/utils/memory/BufferPool.java:
```diff
@@ -148,45 +236,95 @@ public static void putUnusedPortion(ByteBuffer buffer)
+    /**
+     * @return buffer size being used, including used pooled buffers and unpooled buffers
+     */
+    public long usedSizeInBytes()
```
possible alternate naming: `bytesInUse()`

jasonstack replied:
it's to keep consistent with other name `sizeInBytes` and `overflowMemoryInBytes`

maedhroz commented on src/java/org/apache/cassandra/utils/memory/BufferPool.java:
```diff
@@ -54,10 +56,52 @@
+ *     <li>Macro Chunk
+ *       <ul>
+ *         <li>A memory slab that has size of MACRO_CHUNK_SIZE which is 64 * NORMAL_CHUNK_SIZE</li>
+ *         <li>Used to allocat normal chunk with size of NORMAL_CHUNK_SIZE</li>
```
```suggestion
 *         <li>Used to allocate normal chunk with size of NORMAL_CHUNK_SIZE</li>
```

jasonstack replied:
+1

maedhroz commented on src/java/org/apache/cassandra/utils/memory/BufferPool.java:
```diff
@@ -54,10 +56,52 @@
+ *
+ * {@link GlobalPool} maintains two kinds of freed chunks, fully freed chunks where all buffers are released, and
+ * partially freed chunks where some buffers are not released, eg. held by {@link org.apache.cassandra.cache.ChunkCache}.
+ * Partially freed chunks are used to improve cache utilization and has lower priority comparing to fully freed chunk.
```
```suggestion
 * Partially freed chunks are used to improve cache utilization but have lower priority compared to fully freed chunks.
```

maedhroz commented on src/java/org/apache/cassandra/utils/memory/BufferPool.java:
```diff
@@ -288,17 +431,28 @@ private Chunk allocateMoreChunks()
+        public void recyclePartially(Chunk chunk)
+        {
+            if (debug != null)
+                debug.recyclePartial(chunk);
```
It might be worth initializing `debug` with a "nullary" implementation of the `Debug` interface to avoid the `null` checks.

jasonstack replied:
+1

maedhroz commented on src/java/org/apache/cassandra/utils/memory/BufferPool.java:
```diff
@@ -865,10 +1075,19 @@ private static ByteBuffer allocateDirectAligned(int capacity)
+        final long baseAddress;
         private final int shift;
 
+        // it maybe 0L when all slots are allocated after "get" or when all slots are freed after "free"
```
```suggestion
        // it may be 0L when all slots are allocated after "get" or when all slots are freed after "free"
```

maedhroz commented on src/java/org/apache/cassandra/utils/memory/BufferPool.java:
```diff
@@ -878,6 +1097,12 @@ private static ByteBuffer allocateDirectAligned(int capacity)
+                AtomicReferenceFieldUpdater.newUpdater(Chunk.class, Status.class, "status");
+        private volatile Status status = Status.IN_USE;
+
+        private volatile boolean fullyRecyled = false;
```
`fullyRecycled` is unused

maedhroz commented on src/java/org/apache/cassandra/utils/memory/BufferPool.java:
```diff
@@ -878,6 +1097,12 @@ private static ByteBuffer allocateDirectAligned(int capacity)
         private final Recycler recycler;
 
+        private static final AtomicReferenceFieldUpdater<Chunk, Status> statusUpdater =
+                AtomicReferenceFieldUpdater.newUpdater(Chunk.class, Status.class, "status");
```
@jasonstack Using `AtomicReferenceFieldUpdater` here to avoid having tons of `AtomicReference` instances in the `Chunk`s?

jasonstack replied:
yeah, save some memory..

maedhroz commented on src/java/org/apache/cassandra/utils/memory/BufferPool.java:
```diff
@@ -609,23 +782,39 @@ public void put(ByteBuffer buffer, Chunk chunk)
                 chunk.recycle();
             }
-            else if (((free == -1L) && owner != this) && chunk.owner == null)
+            else if (((free == -1L) && owner != this) && chunk.owner == null && !chunk.recycler.canRecyclePartially())
```
nit: lots of unnecessary parentheses here

maedhroz commented on src/java/org/apache/cassandra/utils/memory/BufferPool.java:
```diff
@@ -609,23 +782,39 @@ public void put(ByteBuffer buffer, Chunk chunk)
                 // we must also check after completely freeing if the owner has since been unset, and try to recycle
                 chunk.tryRecycle();
             }
+            else if (chunk.owner == null && chunk.recycler.canRecyclePartially() && chunk.setInUse(Chunk.Status.EVICTED))
```
nit: Do we have to use `chunk.owner` here or will `owner` suffice? (Same for the previous `if`)

jasonstack replied:
it's to avoid race conditions. because the `chunk` can be acquired by another thread after `free`..

maedhroz replied:
Maybe worth documenting here so somebody doesn't mistakenly try to refactor in the future?

maedhroz commented on test/unit/org/apache/cassandra/utils/memory/BufferPoolTest.java:
```diff
@@ -37,6 +38,8 @@
 
 public class BufferPoolTest
 {
+    private final BufferPool bufferPool = BufferPoolManager.permanent();
```
If we just replace this with something like `new BufferPool("Permanent", 512 * 1024 * 1024, true);`, it looks like we can make this a true unit test and get rid of `setupDD()` as well. If fact, why don't we just create a new `BufferPool` for each test in `setUp()` (or in the tests themselves )and get rid of `unsafeSetMemoryUsageThreshold()` *and* make `memoryUsageThreshold` and `readableMemoryUsageThreshold` final? (That would nicely deal with one of my other comments...)

jasonstack replied:
we can do it if we move metrics initialization out of constructor..

maedhroz replied:
It would at least eliminate any concern about the overhead of volatile...

maedhroz replied:
@jasonstack Unless it looks really bad/has some other consequence I'm not foreseeing, it feels like we should attempt this (i.e. to make `memoryUsageThreshold` and `readableMemoryUsageThreshold` final at the expense of a little more test overhead).

jasonstack replied:
pushed 25612b1 to make `memoryUsageThreshold` final..

maedhroz reviewed (commented):
@jasonstack Finished the bulk of my review, although there are still some details of the new tests I want to go over. By and large, this looks pretty good.

maedhroz commented on test/unit/org/apache/cassandra/utils/memory/BufferPoolTest.java:
nit: Perhaps we should actually refer to `NORMAL_CHUNK_SIZE` then?

jasonstack replied:
+1

maedhroz commented on test/unit/org/apache/cassandra/utils/memory/BufferPoolTest.java:
@jasonstack We definitely hit the codepath that partially frees `chunk2`, but we don't ever hit the codepath in `GlobalPool#get()` that polls a chunk from `partiallyFreedChunks`. It seems like we should verify that logic here?

jasonstack replied:
added a verification that: fully freed chunk is prioritized over partially freed chunk

maedhroz commented on test/unit/org/apache/cassandra/utils/memory/BufferPoolTest.java:
```suggestion
        assertSame(partiallyFreed, remainingChunks.get(remainingChunks.size() - 1)); // last one is partial
```

iamaleksey commented:
Remember to make the renames to the two pool names, too, please.

maedhroz commented on src/java/org/apache/cassandra/utils/memory/BufferPool.java:
```diff
@@ -120,8 +120,8 @@
-    private volatile long memoryUsageThreshold;
-    private volatile String readableMemoryUsageThreshold;
+    private final long memoryUsageThreshold;
+    private final String readableMemoryUsageThreshold;
```
+100

jasonstack commented:
@iamaleksey  I have pushed renamed buffer pool commit and also remove overflow-metric update for `putUnUsedPortion`

jasonstack commented:
@maedhroz @iamaleksey I have squashed and rebased.. Also changed config name from `temporary_cache_size_in_mb` to `networking_cache_size_in_mb` to better describe what it does..

maedhroz commented:
@jasonstack Sounds good!
